### PR TITLE
[trainer] Skip the valid-token all-reduce on non-loss PP stages

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Human Note: this remove the big allreduce at the begining of train_step for global_valid_tokens accumulation on all pp ranks, except for the rank that contains last PP stage, i.e. the one contains loss_fn. 

## Summary

`train_step` reduced `global_valid_tokens` across the batch mesh on every
rank. Only the last pipeline stage computes the loss, so every earlier
stage paid a batch-mesh all-reduce for a scalar it never reads:
`pp_forward_backward_step` passes `global_valid_tokens` through
`loss_kwargs`, but the schedule only invokes the loss function where
`target_mbs`/`losses` are populated, which is the last stage alone.

Earlier stages now keep their local count. It stays a device tensor so the
fwd/bwd CUDA graph input signature is unchanged, and their gradients still
arrive normalized by the loss stage's count, since that normalization
happens inside the loss stage's backward.

The skip is uniform across the batch mesh: `pp_has_last_stage` is the same
for every data-parallel rank at a given pipeline stage, so no rank is left
waiting in a collective.

## Numerics

Unchanged on the loss stage, which is the only stage whose
`global_valid_tokens` feeds a loss. On earlier stages the scalar goes
unread by the math; the one observable difference is that
`log_trace_scalar("global_valid_tokens")` reports the rank-local count
there instead of the batch-mesh sum. The metrics block's `local_avg_loss`
also shifts on those stages, but it is already derived from the `-1.0`
sentinel that non-last stages return, and PP loss metrics are only
meaningful on the last stage (see `ensure_pp_loss_visible`).

## Test plan

- `pytest -q tests/unit_tests/cpu/test_trainer.py tests/unit_tests/cpu/test_validate.py
  tests/unit_tests/cpu/test_invalid_loss.py tests/unit_tests/cpu/components` -- 131 passed
- New `test_train_step_reduces_valid_tokens_only_on_loss_stage` covers no-PP,
  PP non-last stage, and PP last stage; it asserts the reduction is skipped and
  the batch mesh is never requested off the loss stage, and that the scalar handed
  to `forward_backward_step` is reduced only where it should be. Verified the test
  fails against the previous unconditional reduction.
- `pre-commit run --files torchtitan/trainer.py tests/unit_tests/cpu/test_trainer.py`

---

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -850,20 +850,24 @@ def train_step(
...
 
-        # Keep the global token count on device so loss normalization does not
-        # introduce a CPU synchronization in the training path.
-        local_valid_tokens_tensor = torch.tensor(
+        # Keep the loss normalization count on device so it does not introduce
+        # a CPU synchronization in the training path. Only the last PP stage
+        # computes the loss, so earlier stages keep their local scalar rather
+        # than reducing one they never use; it still has to be a device tensor
+        # to preserve the fwd/bwd CUDA graph input signature.
+        global_valid_tokens = torch.tensor(
```
Comment (fegin) [inline torchtitan/trainer.py:858]: This one should be local not global. It is pre-dist_sum_tensor result.

File: torchtitan/trainer.py [hunk 2]
```diff
@@ -850,20 +850,24 @@ def train_step(
...
             local_valid_tokens,
             dtype=torch.int64,
             device=self.device,
         )
-        if parallel_dims.dp_enabled:
+        should_reduce_valid_tokens = parallel_dims.dp_enabled and (
```
Comment (tianyu-l) [inline torchtitan/trainer.py:863]: I prefer to not introduce this new variable

File: torchtitan/trainer.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/trainer.py:858]: I think we shouldn't have defined `global_valid_tokens` on non-last-stage because it is not well-defined.

E.g. this logging will produce something "wrong" https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L959

I'd rather leave it `None` and make it optional for PP fwd-bwd (but not spmd fwd-bwd or PP last stage!)
https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L785

Comment (SherlockNoMad) [reply torchtitan/trainer.py:858]: One complexity I ran into is the composability with PP-cudagraph. 

cudagraph wrapper currently takes `global_valid_tokens` as a required input. 

@fegin, when you land pp cudagraph, I can refactor it to support optional tensor input.
